packages/react-components/react-jsx-runtime/src/createElement.ts

```diff
@@ -1,38 +1,60 @@
 import * as React from 'react';
-import { SlotRenderFunction, UnknownSlotProps, SLOT_RENDER_FUNCTION_SYMBOL } from '@fluentui/react-utilities';
-
-type WithMetadata<Props extends {}> = Props & {
-  [SLOT_RENDER_FUNCTION_SYMBOL]: SlotRenderFunction<Props>;
-};
+import {
+  UnknownSlotProps,
+  isSlot,
+  SlotComponent,
+  SLOT_COMPONENT_METADATA_SYMBOL,
+  slot,
+} from '@fluentui/react-utilities';
 
 export function createElement<P extends {}>(
   type: React.ElementType<P>,
   props?: P | null,
   ...children: React.ReactNode[]
 ): React.ReactElement<P> | null {
-  return hasRenderFunction(props)
-    ? createElementFromRenderFunction(type, props, children)
-    : React.createElement(type, props, ...children);
+  // TODO:
+  // this is for backwards compatibility with getSlotsNext
+  // it should be removed once getSlotsNext is obsolete
+  if (isSlot<P>(props)) {
+    return createElementFromSlotComponent(
+      slot(props, { required: true, elementType: type as React.ComponentType<P> }),
+      children,
+    );
+  }
+  if (isSlot<P>(type)) {
+    return createElementFromSlotComponent(type, children);
+  }
+  return React.createElement(type, props, ...children);
 }
 
-function createElementFromRenderFunction<P extends UnknownSlotProps>(
-  type: React.ElementType<P>,
-  props: WithMetadata<P>,
+function createElementFromSlotComponent<Props extends UnknownSlotProps>(
+  slotComponent: SlotComponent<Props>,
   overrideChildren: React.ReactNode[],
-): React.ReactElement<P> | null {
-  const { [SLOT_RENDER_FUNCTION_SYMBOL]: renderFunction, ...renderProps } = props;
+): React.ReactElement<Props> | null {
+  const { [SLOT_COMPONENT_METADATA_SYMBOL]: metadata, as, ...propsWithoutMetadata } = slotComponent;
+  const props = propsWithoutMetadata as UnknownSlotProps as Props;
+  const { elementType: baseElementType, renderFunction } = metadata;
 
-  if (overrideChildren.length > 0) {
-    renderProps.children = React.createElement(React.Fragment, {}, ...overrideChildren);
+  const elementType =
+    baseElementType === undefined || typeof baseElementType === 'string'
+      ? as ?? baseElementType ?? 'div'
```

Comment:
It is an error if the element type is not defined by the time we get here right? Let's not hide the error by defaulting to 'div'. Let it crash if someone managed to get here without setting elementType:
```suggestion
  const elementType =
    typeof baseElementType === 'string'
      ? as ?? baseElementType
```